util/spatial-groups/spatial-groups.lisp

```diff
@@ -0,0 +1,153 @@
+
+;; return to 0,0 on the current taskplane as a shortcut to return to the core task
+(defcommand coord-taskorigin () ()
+  (gselect
+   (group-name
+    (spatial-find-group (current-screen)
+                   (format nil "~{~a~^,~}" (list 0 0 (parse-integer (third (split-string (group-name (current-group)) ",")))))))))
+
+;; pop back to last location
+(defcommand coord-taskpop () () (when *last-cursor* (spatial-gselect *last-cursor*)))
+
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; Key bindings for spatial group movement
+
+(defun install-default-keybinds ()
+  "Install default keybinds. Use this as an example for making your own keybinds."
+
+  ;; Shift arrows between adjacent frames on the current screen
+  (define-key *top-map* (kbd "S-Up")      "move-focus up")
+  (define-key *top-map* (kbd "S-Down")    "move-focus down")
+  (define-key *top-map* (kbd "S-Left")    "move-focus left")
+  (define-key *top-map* (kbd "S-Right")   "move-focus right")
+
+  ;; Control arrows move between screens on the current desktop
+  (define-key *top-map* (kbd "C-Left")    "coord-left")
+  (define-key *top-map* (kbd "C-Right")   "coord-right")
+  (define-key *top-map* (kbd "C-Up")      "coord-up")
+  (define-key *top-map* (kbd "C-Down")    "coord-down")
+
+  ;; Control-Shift left/right to switch desktop Z
+  (define-key *top-map* (kbd "C-S-Left")  "coord-taskleft")
+  (define-key *top-map* (kbd "C-S-Right") "coord-taskright")
+
+  ;; Control-Shift-Up to return to origin 0,0 on current desktop Z
+  (define-key *top-map* (kbd "C-S-Up")    "coord-taskorigin")
+
+  ;; "Pop" back to last desktop position
+  (define-key *top-map* (kbd "C-S-Down")  "coord-taskpop") )
+
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; Spatial groups and coordinate functions
+
+;; Groups will manage the coordinate system
+;; format: 0,0,0 with positive and negative numbers
+;; create groups on the fly as needed
+;; only supports one screen atm, would like multimonitor support later
+
+(defun spatial-find-group (screen name)
+  "Simple function to find group by string name"
+  (find name (screen-groups screen) :key 'group-name :test 'string=))
+
+(defun spatial-gselect (name)
+  "Wrap gselect, preserve prior location for pop, and handle when group is new"
+  (when *spatial-banish-on-move* (banish))
+  (setf *last-cursor* (group-name (current-group)))
+  (gselect (group-name
+             (or (spatial-find-group (current-screen) name)
+               (gnew name)))))
+
+(defun coord-group-change (xo yo zo)
+  "Navigate a 3d array of groups using x,y,z coordinates by passing the offset of the change."
+  (let* ((current-coords
+          (mapcar #'parse-integer
+                  (split-string (group-name (current-group)) ",")))
+         (new-coords (mapcar #'+ current-coords (list xo yo zo)))
+         (new-group-name (format nil "~{~a~^,~}" new-coords)) )
+
+    (if (= 0 zo)
```

Comment:
some one mentioned if-let, which I am not deeply familiar with, however, I will second my general feeling that "long if statements are hard to read, and make me think harder than I'd generally like